libportal/notification.c

```diff
@@ -19,9 +19,79 @@
 
 #include "config.h"
 
+#include <fcntl.h>
+
+#include <glib/gstdio.h>
+#include <gio/gunixfdlist.h>
+
 #include "notification.h"
 #include "portal-private.h"
 
+static void
+prepare_media (GVariantBuilder *builder,
+              GVariant *serialized_media,
+              GUnixFDList **fd_list)
+{
+  g_autoptr(GVariant) value = NULL;
+  g_autoptr(GFile) file = NULL;
+  const char *key;
+  g_autofree char *path = NULL;
+  int fd, fd_in = 0;
+
+  if (g_variant_is_of_type (serialized_media, G_VARIANT_TYPE("(sv)")))
+    g_variant_get (serialized_media, "(&sv)", &key, &value);
+
+  if (key && value && strcmp (key, "file") == 0 && g_variant_is_of_type (value, G_VARIANT_TYPE_STRING))
+    file = g_file_new_for_commandline_arg (g_variant_get_string (value, NULL));
+
+  /* Pass through everything that isn't a native GFile */
+  if (!G_IS_FILE (file) || !g_file_is_native (file))
+    {
+      g_variant_builder_add_value (builder, serialized_media);
+      return;
+    }
+
+  path = g_file_get_path (file);
+  fd = g_open (path, O_RDONLY | O_CLOEXEC);
+
+  if (*fd_list == NULL)
+    *fd_list = g_unix_fd_list_new_from_array (&fd, 1);
+  else
+    fd_in = g_unix_fd_list_append (*fd_list, fd, NULL);
```

Comment:
That seems like a really roundabout way of handling things. Why not just `g_unix_fd_list_new` in the caller and pass it through to here?